### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [pemistahl/lingua-rs#273] [Ideas]
Hi!

Writing this for the history. Maybe these results will be interesting to someone who trying to achieve better performance with `lingua-rs` since the project [cares](https://www.reddit.com/r/rust/comments/17vrnpg/lingua_160_the_most_accurate_natural_language/) about performance.

I test Profile-Guided Optimization (PGO) on different kinds of software - the current results are available [here](https://github.com/zamazan4ik/awesome-pgo)(with a lot of other PGO-related information). That's why I tried to optimize `lingua-rs` with PGO too.

## Test environment

I performed tests on two configurations: Linux and macOS.

Linux:

* Fedora 39
* Linux kernel 6.5.11
* AMD Ryzen 9 5900x
* 48 Gib RAM
* SSD Samsung 980 Pro 2 Tib
* Compiler - Rustc 1.73
* Lingua-rs version: the latest for now from the `main` branch on commit `5a1f4980162996319474552e319b211035fdd670`
* Disabled Turbo boost

## Benchmarks

As a benchmark, I use built-in benchmarks with `cargo bench --all-features`. For PGO optimization phase, I use [cargo-pgo](https://github.com/Kobzol/cargo-pgo) with `cargo pgo optimize bench -- --all-features`. For PGO training phase, I use the same benchmark with `cargo pgo bench -- --all-features`.

## Results

I got the following results:

* PGO-optimized compared to Release: https://gist.github.com/zamazan4ik/fdbb9f8b2a1e16d6bbd162a01ea0f6b5
* (just for reference) PGO-instrumented compared to Release: https://gist.github.com/zamazan4ik/14973a900c780902462a29cccbe18ae8

As you see, in several cases lingua-rs performance can be improved with PGO. I think this information can be written somewhere into the documentation, so users will be aware about PGO effects on Lingua-rs performance and can decide to apply PGO for their Lingua-rs builds.

**Comment by pemistahl** [comment]
Hi, sorry for the late response and thanks for this valuable information. I'm sure there is still room for performance improvements. As soon as I want to work on performance, I will look at your results more closely.
